### ISSUE TITLE
Support signatures

### ISSUE THREAD

**Issue by MarcSkovMadsen** [microsoft/pyright#2855]
Hi Pyright

I am a contributor to and user of the [HoloViz](https://holoviz.org/) ecosystem. Especially [Panel](https://panel.holoviz.org/). I develop in VS Code. The problem is I don't get much context help. This is an example

![image](https://user-images.githubusercontent.com/42288570/149266449-c7d5b416-8df5-4d64-a54b-d34efcd737bd.png)

The HoloViz ecosystem is built on top of [Param](https://param.holoviz.org/) which provides powerful parameters for your classes. They are especially powerful for building GUIs. They support events, watchers etc. There is some overlap or similarity to dataclasses, attrs, traits, traitlets, pydantic etc.

The implementation of Param uses meta classes and a lot of inheritance.

@philippjfr points out in https://github.com/holoviz/panel/issues/3085 that *..., we explicitly enabled [signatures](https://www.python.org/dev/peps/pep-0362/) (pep 362) on Parameterized class constructors which should be the accepted way to declare the accepted kwargs.*

@brettcannon points out in https://github.com/holoviz/panel/issues/3085 that *For attrs at least, they implemented https://github.com/microsoft/pyright/blob/main/specs/dataclass_transforms.md which I believe is in the process of becoming a proposed PEP. It's documented at https://www.attrs.org/en/stable/extending.html?highlight=pyright#pyright .*

## Questions

- Is it correct that Pyright does not use `signatures` (pep 362)?
- Why?
- Could you support it?
- Would you support it?

Alternatively we (Param) would have to implement the `dataclass_transform`. I've made an FR here https://github.com/holoviz/param/issues/587. Feel free to join the discussion. 

Any input is appreciated. Thanks

## Example of VS Code use

![autoreload (2)](https://user-images.githubusercontent.com/42288570/149266319-785c5425-1f94-437d-a4d3-9e80c34e6e64.gif)

**Comment by erictraut** [comment]
Pyright is a static type checker. It doesn't run any Python code. Running arbitrary code would be a security issue and a performance problem. As such, it uses only statically-available type information. If I understand "signatures" correctly, it is based on runtime introspection of function objects. That won't work with static type checking. If you're interested in getting completion suggestions or other type checking functionality, you should annotate your function and method signatures with type annotations as specified in PEP 484, etc.

**Comment by erictraut** [comment]
If your library is using dataclass-like semantics (similar to attrs or pydantic), then the proposed `dataclass_transform` mechanism might be a good answer. We recently posted a draft PEP and are actively soliciting feedback and input on [this typing-sig thread](https://mail.python.org/archives/list/typing-sig@python.org/thread/EAALIHA3XEDFDNG2NRXTI3ERFPAD65Z4/).

**Comment by erictraut** [comment]
You may also find [this documentation](https://github.com/python/typing/blob/master/docs/source/libraries.rst) useful. I wrote this guidance for library authors and maintainers. Let me know if you have any follow-up questions.

**Comment by erictraut** [comment]
Looking more closely at `panels`, I don't think that `dataclass_transform` will work for you. The classes are not close enough to dataclass semantics.

The best suggestion I can offer is to implement annotated `__init__` methods in each of the classes. The signature should explicitly list and annotate the accepted keyword parameters for each class's constructor. Ideally, a docstring would also be provided for each class too. For example:

```python
class IntSlider(ContinuousSlider):
    def __init__(
        self,
        *,
        value: int = 0,
        value_throttled: bool = True,
        start: int = 0,
        end: int = 1,
        step: int = 1,
    ):
        """A continuous slider with integer start, stop and step values."""
        super().__init__(
            value=value,
            value_throttled=value_throttled,
            start=start,
            end=end,
            step=step,
        )
```

Given that you have good introspection capabilities, you could probably automate the creation of these `__init__` methods.

Here's what pyright & pylance do with this information.

![Screen Shot 2022-01-12 at 10 04 01 PM](https://user-images.githubusercontent.com/7040122/149274661-47c7b5e4-f71b-4835-a201-4433e6561fe8.png)

**Comment by MarcSkovMadsen** [comment]
Thanks so much @erictraut. That is really useful.

**Comment by philippjfr** [comment]
Thanks for getting this discussion started everyone! I'd be very interested in implementing support for this, currently we indeed generate the constructor signatures at run-time, which I understand is pretty much worthless for pyright. I'm not excited about repeating all parameter declarations in the constructor of each class (the baseclass for all Panel objects alone has 14 parameters). The only feasible way I can imagine supporting this at the moment would be to insert the signatures during a build step, i.e. actually transforming the source code before a release so the packaged library includes the signatures. If anyone has an alternative suggestion I'd love to hear it.

**Comment by MarcSkovMadsen** [comment]
The alternative would probably be refactoring Param to align with `dataclasses`. It would probably be as troublesome as moving from Python2 to Python3 for developers and users. But it would modernize everything and help every library built on top of Param in the future.

Or alternatively replace Param with some library that is aligned with dataclasses and supports events.

**Comment by MarcSkovMadsen** [comment]
I guess stubs are also a possibility. I don't know how they work. But probably it would be safer/ better to autogenerate these during build than insert into the code ??

**Comment by erictraut** [comment]
Type stubs are an option. You could programmatically generate ".pyi" files and ship them alongside your ".py" files. Type checkers will prefer ".pyi" files if both are present.

**Comment by philippjfr** [comment]
>Type stubs are an option. You could programmatically generate ".pyi" files and ship them alongside your ".py" files. Type checkers will prefer ".pyi" files if both are present.

Thanks, this sounds like a great approach. That's probably what we'll do!
